cpp/include/raft/core/resources.hpp

```diff
@@ -0,0 +1,107 @@
+ * to keep its usage somewhat opaque to end-users.
+ *
+ * @code{.cpp}
+ * #include <raft/core/resources.hpp>
+ * #include <raft/core/resource/cuda_stream.hpp>
+ * #include <raft/core/resource/cublas_handle.hpp>
+ *
+ * raft::resources res;
+ * auto stream = raft::resource::get_cuda_stream(res);
+ * auto cublas_handle = raft::resource::get_cublas_handle(res);
+ * @endcode
+ */
+class resources {
+ public:
+  resources() {}
+
+  resources(const resources&) = delete;
+  resources& operator=(const resources&) = delete;
+  resources(resources&&)                 = delete;
+  resources& operator=(resources&&) = delete;
+
+  /**
+   * @brief Returns true if a resource_factory has been registered for the
+   * given resource_type, false otherwise.
+   * @param resource_type resource type to check
+   * @return true if resource_factory is registered for the given resource_type
+   */
+  bool has_resource_factory(resource::resource_type resource_type) const
+  {
+    std::lock_guard<std::mutex> _(mutex_);
+    return factories_.find(resource_type) != factories_.end();
+  }
+
+  /**
+   * @brief Register a resource_factory with the current instance.
+   * This will overwrite any existing resource factories.
+   * @param factory resource factory to register on the current instance
+   */
+  void add_resource_factory(std::shared_ptr<resource::resource_factory> factory) const
+  {
+    std::lock_guard<std::mutex> _(mutex_);
+    factories_.insert(std::make_pair(factory.get()->get_resource_type(), factory));
+  }
+
+  /**
+   * @brief Retrieve a resource for the given resource_type and cast to given pointer type.
+   * Note that the resources are loaded lazily on-demand and resources which don't yet
+   * exist on the current instance will be created using the corresponding factory, if
+   * it exists.
+   * @tparam res_t pointer type for which retrieved resource will be casted
+   * @param resource_type resource type to retrieve
+   * @return the given resource, if it exists.
+   */
+  template <typename res_t>
+  res_t* get_resource(resource::resource_type resource_type) const
+  {
+    std::lock_guard<std::mutex> _(mutex_);
+    if (resources_.find(resource_type) == resources_.end()) {
+      RAFT_EXPECTS(factories_.find(resource_type) != factories_.end(),
+                   "No resource factory has been registered for the given resource %d.",
+                   resource_type);
+      resource::resource_factory* factory = factories_.at(resource_type).get();
+      resources_.insert(std::make_pair(resource_type, factory->make_resource()));
+    }
+    return reinterpret_cast<res_t*>(resources_.at(resource_type).get()->get_resource());
+  }
+
+ private:
+  mutable std::mutex mutex_;
+  mutable std::unordered_map<resource::resource_type, std::shared_ptr<resource::resource_factory>>
+    factories_;
+  mutable std::unordered_map<resource::resource_type, std::shared_ptr<resource::resource>>
```

Comment:
Convert the `shared_ptr` to a `unique_ptr` since we never return the `shared_ptr` to the user, just the raw pointer. That means the lifetime of this resource is entirely managed by the creating container of this resource, and thus no extra bookkeeping needs to be done